Plotting for ZXDiagrams and ZXGraphs, Equality for ZXDiagrams and converting ZX to ZXW

contra-bit opened the pull request:
Dear Maintainers,

This pull request introduces several improvements, including the addition of optional plotting functionality for  ZXDiagrams and ZXGraphs and closes #101 and fixes #100. The plotting function  uses Vega and DataFrames, which will only be loaded when necessary to improve load times and reduce memory usage. The color scheme of spider graphs has been updated to be more accessible for individuals with Color Vision Deficiency (CVD) and improved label readability. Further information on accessibility can be found at <[https://zxcalculus.com>](https://zxcalculus.com).


Additional testing has been implemented for both ZXDiagrams and ZXGraphs, with plots added to most tests and the creation of a new notebook focused on equivalence checking. Existing tutorial notebooks have also been refined, and users will now benefit from automated diagram generation within these notebooks.

Furthermore, the YaoHIR package has gained the capability to import OpenQASM files, enabling easy conversion from openQASM to ZXDiagrams and providing another useful approach to building. optimizing and verifying quantum circuits. 

The pull request further includes various bug fixes and improvements, such as addressing issues with the match and apply function for the ZXGraph :id rule, enhancing the overall user experience for plots, resolving testing issues, and handling miscellaneous bugs.

Regards and thanks for reviewing my pull request

exAClior commented:
Hello @contra-bit , thank you very much for working on adding `plot` into `ZXCalculus.jl`.  I was trying out the notebook `tutorial.jl`. I was able to see the new plots, they looked cool! But were you able to install all the required packages that are at the beginning of this notebook? I could not do it. I noticed that you added `CompilerPluginTools` to dependency. That was causing a little trouble for us. It will probably cause more trouble for us when `Julia v1.10` is out. Roger is planning on updating YaoHIR and removing this completely see [link](https://github.com/JuliaCompilerPlugins/CompilerPluginTools.jl/pull/10#issuecomment-1619228573). Is this a hard requirement for your future work regarding this dependency?

contra-bit commented:
Good catch. It is not required.

I only added `CompilerPluginTools` as an explicit dependency because it was causing trouble for me, when it was being  imported as a dependency by YaoHIR. 
The import failed because the version required could not be resolved.
The only way for me to resolve the `YaoHIR` dependency was to add the master branch from the git repo because the julia package had the wrong version. 
In order for it to be imported correctly I ran the following command:

```julia
pkg> add https://github.com/JuliaCompilerPlugins/CompilerPluginTools.jl
```
Then I added the equivalent command to the notebook as well. 

```julia
pkg> add https://github.com/JuliaCompilerPlugins/CompilerPluginTools.jl
```

ChenZhao44 commented:
@contra-bit Thanks for the PR. I suggest removing the dependency that is only used in the notebook, and keeping the main package as clean as possible. Perhaps it would be better to set up a separate environment in the `notebook` folder to include `CompilerPluginTools` and `DataFrames`.

contra-bit commented:
I am thrilled to share several improvements in this pull request! I have introduced plotting functionality as an optional extension using Vega and DataFrames. This change ensures that these dependencies will only be loaded when needed, leading to faster load times and reduced memory usage.

To enhance accessibility, I have updated the color scheme of spider graphs so that they are distinguishable for people with Color Vision Deficiency (CVD), and improved the readability of phase labels. For further details, kindly consult the accessibility guidelines provided at zxcalculus.com.

My latest contributions enhance testing for both ZXDiagrams and ZXGraphs by adding plots to most tests, along with creating a brand-new notebook dedicated to equivalence checking and refining the existing tutorial notebook. Users will benefit from automatic generation of diagrams within these notebooks.

Moreover, YaoHIR has gained the ability to import OpenQASM files, allowing seamless conversion to ZXDiagrams—providing yet another practical method to construct quantum circuits. I will try to get all my contributions merged and then update the Pkg in the notebooks.

As part of my efforts, I fixed the match and apply function for the  ZXGraph :id rule while adding the new equivalence checking functionality. Now, it is possible to determine if two diagrams are identical through the built-in equivalency checks.

Last but not least, I took care of numerous housekeeping tasks such as updating imports, resolving problems discovered during testing, enhancing overall user experience for plots, and dealing with miscellaneous glitches

ChenZhao44 commented:
@contra-bit Thanks for working on this PR. It looks decent. Before merging this PR, I have only two minor suggestions.
- You don't need to keep the original `tutorial.jl` notebook since it is out-of-date.
- There are a lot of indent changes in this PR. The default indent of this repo is 4, but some of the indents are set to 2. These changes have created a lot of unnecessary differences, making it difficult to keep track of the real changes. Could please you fix the indent issue?

Roger-luo commented:
FYI you don't have to use the package env for notebooks, every Pluto notebook has its own env by default.

contra-bit commented:
I hope I have fixed the indentation issues, further I have increased the test coverage and added a .JuliaFormatter.toml that species that the default indentation is 4.

The notebooks are now configured to use the Yao upstream repositories. 
My next goal is to improve the notebooks and improve the documentation.

Roger-luo commented on notebooks/zx_intro.jl:
```diff
@@ -11,8 +11,8 @@ begin
+    Pkg.add("ZXCalculus")
     Pkg.add(
         url = "https://gitlab.informatik.uni-bremen.de/qce/quantumequivalencechecker.jl",
         rev = "feat/zx",
```
can we use Pluto builtin env instead of doing this?

contra-bit replied:
Sure

ChenZhao44 commented on src/ZX/rules.jl:
```diff
@@ -896,6 +909,11 @@ function check_rule(::Rule{:id}, zxg::ZXGraph{T, P}, vs::Vector{T}) where {T, P}
+                  return true
+                end
+
             else phase(zxg, v2) == 1
```
should also be `elseif phase(zxg, v2) == 1`

ChenZhao44 commented on src/ZX/rules.jl:
```diff
@@ -912,19 +930,26 @@ function rewrite!(::Rule{:id}, zxg::ZXGraph{T, P}, vs::Vector{T}) where {T, P}
+      if ((spider_type(zxg, v1) == SpiderType.In || spider_type(zxg, v1) == SpiderType.Out) && 
+          (spider_type(zxg, v3) == SpiderType.In || spider_type(zxg, v3) == SpiderType.Out))
+          rem_spider!(zxg, v2)
+          add_edge!(zxg, v1, v3, EdgeType.SIM)
```
Do you need to check the edge type for the edge (v1, v2) and (v2, v3)? For example, if only one of them are Hadamard edge, then the resulting edge should also be a Hadamard edge.

contra-bit replied:
Your assumption seems to be correct. I replaced the and with and or

ChenZhao44 commented on src/ZXW/utils.jl:
```diff
@@ -667,3 +667,14 @@ function substitute_variables!(
+
+
+function convert_to_zxwd(root::YaoHIR.BlockIR)
+    # FIXME add test
```
add tests?

contra-bit replied:
Test is in ir.jl in line 57. Forgot to remove the FIXME

ChenZhao44 commented on test/ancilla_extraction.jl:
```diff
@@ -17,14 +17,14 @@ zxd = gen_phase_gadget()
 anc_circ = ancilla_extraction(zxg)
 
-# plot(anc_circ)
+@test plot(anc_circ) !== nothing
```
better to use `isnothing`

ChenZhao44 commented on test/ancilla_extraction.jl:
```diff
@@ -17,14 +17,14 @@ zxd = gen_phase_gadget()
 zxg_swap = ZXGraph(zxd_swap)
 zxd_anc = ancilla_extraction(zxg_swap)
-# plot(zxd_anc)
+@test plot(zxd_anc)  !== nothing
```
use `isnothing`

ChenZhao44 commented on test/challenge.jl:
```diff
@@ -208,4 +208,5 @@ for i = 1:5
     push!(zxg.outputs, i+23)
 end
 
+@test plot(zxg) !== nothing
```
use `isnothing`

ChenZhao44 commented on test/circuit_extraction.jl:
```diff
@@ -38,3 +38,9 @@ zxg2 = clifford_simplification(zxd)
 cir = circuit_extraction(zxg3)
+
+# Plotting
+@test plot(zxd) !== nothing
```
use `isnothing`

ChenZhao44 commented on test/phase_teleportation.jl:
```diff
@@ -75,6 +75,8 @@ function gen_cir()
 end
 
 cir = gen_cir()
+@test plot(cir)!== nothing
```
use `isnothing`

ChenZhao44 commented on test/plots.jl:
```diff
@@ -0,0 +1,13 @@
+ZX.insert_spider!(zxd, 1, 2, SpiderType.Z)
+zxg = ZXGraph(zxd)
+
+@test plot(zxd) !== nothing
```
use `isnothing`

ChenZhao44 commented on test/rules.jl:
```diff
@@ -10,6 +10,7 @@ matches = match(Rule{:f}(), zxd)
 rewrite!(Rule{:f}(), zxd, matches)
 @test sort!(spiders(zxd)) == [1, 3]
 @test phase(zxd, 1) == phase(zxd, 3) == 3 // 4
+@test plot(zxd) !== nothing
```
use `isnothing`

ChenZhao44 commented on test/zx_diagram.jl:
```diff
@@ -7,11 +7,13 @@ v_t = [SpiderType.X, SpiderType.Z, SpiderType.X]
 zxd = ZXDiagram(g, v_t, ps)
 zxd2 = ZXDiagram(g, Dict(zip(1:3, v_t)), Dict(zip(1:3, ps)))
 @test zxd.mg == zxd2.mg && zxd.st == zxd2.st && zxd.ps == zxd2.ps
+@test plot(zxd) !== nothing
```
use `isnothing`

contra-bit commented:
I have addressed the feedback and think this can be merged now. @ChenZhao44 

Thank you for being patient.

exAClior commented on src/ZX/circuit_extraction.jl:
```diff
@@ -285,7 +285,7 @@ function update_frontier!(zxg::ZXGraph{T, P}, gads::Set{T}, frontier::Vector{T},
         nb_v = neighbors(zxg, v)
         u = findfirst([u in gads for u in nb_v])
-        if u !== nothing
+          if !isnothing(u)
```
indentation is still messup, could you please check your formatting script? did it include this directory?

exAClior commented on src/ZX/equality.jl:
```diff
@@ -0,0 +1,19 @@
+
+"""
+function verify_equality(zxd_1::ZXDiagram, zxd_2::ZXDiagram)
+  merged_diagram = concat!(zxd_1, dagger(zxd_2))
```
same thing here, the indentation is messed up

exAClior commented on src/ZX/zx_diagram.jl:
```diff
@@ -629,3 +638,180 @@ function safe_convert(::Type{T}, x::Real) where T<:Rational
+
+Add non input and output spiders of d2 to d1, modify d1. Record the mapping of vertex indices.
+"""
+function import_non_in_out!(
```
indentation problem here

exAClior commented on src/ZX/zx_graph.jl:
```diff
@@ -299,7 +302,7 @@ function spider_sequence(zxg::ZXGraph{T, P}) where {T, P}
         end
         for v in vs
-            if qubit_loc(zxg, v) !== nothing
+          if !isnothing(qubit_loc(zxg, v))
```
indentation here

exAClior commented on test/runtests.jl:
```diff
@@ -1,4 +1,15 @@
+
+
+@testset "plots.jl" begin
+  include("plots.jl")
```
indentation here

exAClior reviewed (commented):
Hello @contra-bit , sorry for jumpping in. I rechecked your commits based on @ChenZhao44  and @Roger-luo 's comments. There are still a few outstanding problems.

contra-bit commented on .github/workflows/Documentation.yml:
I missed that it was in CI file. I have removed it

contra-bit commented:
I reformatted the mentioned files by hand. Thank you for the thorough review.

exAClior commented:
Great! This commit looks good to me. Let's see if Dr. Zhao and Dr. Luo has any further comments!

Roger-luo commented:
LGTM
